dpe/src/commands/derive_child.rs

```diff
@@ -64,27 +64,86 @@ impl DeriveChildCmd {
-    /// differentiate between them.
+    /// When a default context is in a locality, it MUST be the only context in the locality. This
+    /// checks that the operation will result in the locality will only have the newly created
+    /// default context and no others.
     ///
     /// # Arguments
     ///
     /// * `parent_idx` - Index of the soon-to-be parent.
     /// * `default_context_idx` - Index of the current default context, if there is one.
+    /// * `num_contexts_in_locality` - Number of contexts already in the locality.
     fn safe_to_make_default(
         &self,
         parent_idx: usize,
-        default_context_idx: Result<usize, DpeErrorCode>,
+        default_context_idx: Option<usize>,
+        num_contexts_in_locality: usize,
     ) -> bool {
-        if let Ok(default_idx) = default_context_idx {
-            if default_idx != parent_idx || self.retains_parent() {
-                return false;
+        match (num_contexts_in_locality, default_context_idx) {
+            // No other contexts in the locality.
+            (0, None) => true,
+            // There is only one context, but that context is the parent.
+            (1, Some(default_idx)) if default_idx == parent_idx => {
+                // As long as the parent is about to be retired, it is okay.
+                !self.retains_parent()
             }
+            // In all other scenarios, there will be a combination of default and non-default
+            // contexts
+            _ => false,
         }
-        true
+    }
+
+    /// Whether it is okay to make a NON-default context.
+    ///
+    /// There can never be a mixture of default and non-default contexts. This checks to make sure
+    /// the operation will not result in a mixture.
+    ///
+    /// # Arguments
+    ///
+    /// * `parent_idx` - Index of the soon-to-be parent.
+    /// * `default_context_idx` - Index of the current default context, if there is one.
+    fn safe_to_make_non_default(
+        &self,
+        parent_idx: usize,
+        default_context_idx: Option<usize>,
+    ) -> bool {
+        match default_context_idx {
+            None => true,
+            // If the default context is the parent, it will either be retired or made into a
+            // non-default context.
+            Some(default_idx) => default_idx == parent_idx,
+        }
+    }
+
+    /// Whether it is okay to create a child in the given environment.
+    ///
+    /// When a default context is in a locality, it MUST be the only context in the locality. There
+    /// can never be a mixture of default and non-default contexts. This checks that the operation
+    /// will not violate either statement.
+    ///
+    /// # Arguments
+    ///
+    /// * `parent_idx` - Index of the soon-to-be parent.
+    /// * `default_context_idx` - Index of the current default context, if there is one.
+    /// * `num_contexts_in_locality` - Number of contexts already in the locality.
+    fn safe_to_make_child(
```

Comment:
nit: There is another piece of code below that is also relevant to whether it's safe to create the child:

```rust
        // [jhand editorial] This statement is only hit if parent_idx handle is default
        } else if locality == target_locality {
            // The default context cannot be retained because a locality cannot
            // support the default context and random context handles at the same time.
            return Err(DpeErrorCode::InvalidArgument);
        }
```

I'm wondering if it would make sense to move that here so that all of these cases are handled in one place. But it is a little different (it's about whether it's safe to retain the parent). So maybe it's ok to handle it separately.

Reply:
I was picturing that the parent context always rotated, even if the parent was a default. Fixed logic and removed the section mentioned.